### PR TITLE
expression: fix constItem() and column-pruning in `order by` clause

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
https://github.com/pingcap/tidb/pull/10064 Change the semantics of constItem().
A const item can be eval() when build a plan, and it shouldn't contain `correlated column`.
 
Column-pruning in `order by` clause, however, need to identify a "const item" in execute stage. And `IsMutableEffectsExpr` should be enough.

### What is changed and how it works?
Change back constItem() and use `IsMutableEffectsExpr`. And export `isMutableEffectsExpr` as `IsMutableEffectsExpr`

### Check List 

Tests 

 - Unit test
Pass the test add by #10064.

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by wjhuang2016** [comment]
@lysu @zz-jason PTAL

File: expression/expression.go [hunk 1]
```diff
@@ -105,6 +105,7 @@ type Expression interface {
 	IsCorrelated() bool
 
 	// ConstItem checks if this expression is constant item, regardless of query evaluation state.
+	// A constant item can be eval() when build a plan.
```
Comment (qw4990) [inline expression/expression.go]:
```suggestion
	// A constant item can be eval() when building a plan.
```

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -154,7 +154,7 @@ func (ls *LogicalSort) PruneColumns(parentUsedCols []*expression.Column) error {
 	for i := len(ls.ByItems) - 1; i >= 0; i-- {
 		cols := expression.ExtractColumns(ls.ByItems[i].Expr)
 		if len(cols) == 0 {
-			if !ls.ByItems[i].Expr.ConstItem() {
+			if expression.IsMutableEffectsExpr(ls.ByItems[i].Expr) {
```
Comment (qw4990) [inline planner/core/rule_column_pruning.go]: `CorCols` will be regarded as immutable by this function and then they will be pruned.
Is it OK to do that?

Comment (wjhuang2016) [reply planner/core/rule_column_pruning.go]: It's expected behavior. Order by corcols is useless and can be pruned.

Comment (qw4990) [reply planner/core/rule_column_pruning.go]: OK, and please add some comments here about it.

Comment (wjhuang2016) [reply planner/core/rule_column_pruning.go]: done!

**Comment by wjhuang2016** [comment]
> Can you add an explain test for this PR? thanks

Done

File: executor/sort_test.go [hunk 3]
```diff
@@ -44,3 +44,22 @@ func (s *testSuite4) TestSortRand(c *C) {
...
+
+func (s *testSuite4) TestSortCorrelatedColumn(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1(a int, b int);")
+	tk.MustExec("drop table if exists t2;")
+	tk.MustExec("create table t2(a int, b int);")
+
+	tk.MustQuery("explain select * from t1 where t1.a in (select t2.a as a from t2 where t2.b > t1.b order by t1.b);").Check(testkit.Rows(
```
Comment (zz-jason) [inline executor/sort_test.go]: it's better to move this test to `explain-test`

Comment (wjhuang2016) [reply executor/sort_test.go]: Would you mean all the tests in this file or just this test?

File: executor/explain_test.go [hunk 4]
```diff
@@ -152,3 +152,49 @@ func (s *testSuite1) checkMemoryInfo(c *C, tk *testkit.TestKit, sql string) {
 		}
 	}
 }
+
+func (s *testSuite4) TestExplainSortRand(c *C) {
```
Comment (eurekaka) [inline executor/explain_test.go]: Better to put them in `cmd/explaintest/`?

**Pull request merged**